README
                              jpgtn version 2
                              ===============
 
  What is jpgtn?
  --------------
 
  Jpgtn is a thumbnailer for JPEG images. 

  Jpgtn is based on "tnpic" by Russell Marks and on "gtnpic" by Willie
  Daniel. "tnpic" was originally distributed as a utility with Marks' "zgv"
  program. "gtnpic" was an attempt to clean up and expand upon the original 
  tnpic program and "maybe add some new features along the way." Jpgtn 
  differs from both tnpic and gtnpic in purpose. It is not an indexer.

  Jpgtn only thumbnails JPEG images. I have no plans to change it to accept 
  other input formats. On and off I've thought about adding support for PNG 
  output but I haven't had a real need for it myself so I haven't done it. 

  Jpgtn can output thumbnails to files with names based on the names of the 
  input files, or it can dump to stdout. The stdout support can be a useful
  feature for CGI applications.

  Jpgtn is based in part on the work of the Independent JPEG Group.



  What do I need to use it?
  -------------------------

  In order to compile the source distribution you will need a C compiler. 

  You will also need the Independent JPEG Group's JPEG library, libjpeg. 
  It is a common library and it may already be installed on your system. 
  If you don't have libjpeg, it is freely available and links to the source 
  and documentation can be found on the Independent JPEG Group's page at 
  http://www.ijg.org/. 



  Compiling
  ---------

  Please see the INSTALL file!


  Contact Info
  ------------

  Email: jpgtn@madea.net
  Web:   http://madea.net/projects/jpgtn



-----

Copyright 2002-2012 Jeremy Madea

This file is part of jpgtn.

Jpgtn is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2, or (at your option)
any later version.

Jpgtn is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with jpgtn; see the file COPYING.  If not, write to
the Free Software Foundation, Inc., 59 Temple Place - Suite 330,
Boston, MA 02111-1307, USA.


NEWS
New in 2.06: 
* Man page is now installed correctly.

New in 2.05: 
* No changes seen by the end user.

New in 2.04:
* No changes seen by the end user. 

New in 2.03:
* --with-libjpeg=PATH configure option. 

New in 2.02:
* GNU autoconf and automake driven build process
* The directory structure for the package has changed.


src/jpgtn.h
/*
    Copyright (C) 2002-2012 Jeremy Madea <jpgtn@madea.net>

    This file is part of jpgtn.

    Jpgtn is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

    Jpgtn is based on two programs: tnpic (distributed with zgv) and gtnpic.
    The original "gtnpic" program was Copyright (C) 1997      Willie Daniel. 
    The original "tnpic"  program was Copyright (C) 1993-1996 Russell Marks.
    Both tnpic and gtnpic were distributed under the GNU Public License. 
*/

#ifndef JPGTN_H
#define JPGTN_H 1

#define MAXSIZE     2048                    /* Maximum size of thumbnail   */
#define MINSIZE     1                       /* Minimum size of thumbnail   */
#define DEFAULTSIZE 128                     /* Default size of thumbnail   */

#define GRAY 128

#define RSZ_AUTO   0
#define RSZ_WIDTH  1
#define RSZ_HEIGHT 2

unsigned char *outimage;                    /* The current thumbnail image */
unsigned char *palette;                     /* Global palette pointer      */
int            width;                       /* Original image width        */
int            height;                      /* Original image height       */
int            background;                  /* Index's background color    */
long           out_wide;                    /* Width of thumbnail image    */
long           out_high;                    /* Height of thumbnail image   */


/* resizepic.c */
extern unsigned char *resizepic(unsigned char *thisimage, 
                                unsigned char *palr,
                                unsigned char *palg, 
                                unsigned char *palb,
                                int            currwidth, 
                                int            currheight,
                                int            maxsize,
                                int            which_dim);

/* readjpeg.c */
extern unsigned char *read_JPEG_file(char *filename, unsigned char **palette);

/* writejpeg.c  */
extern void write_JPEG_file(char *filename, 
                            int   outwidth,  
                            int   outheight, 
                            int   quality);

#endif /* !JPGTN_H */


src/readjpeg.c
/*
    Copyright (C) 2002-2012 Jeremy Madea <jpgtn@madea.net>

    This file is part of jpgtn.

    Jpgtn is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

    Jpgtn is based on two programs: tnpic (distributed with zgv) and gtnpic.
    The original "gtnpic" program was Copyright (C) 1997      Willie Daniel. 
    The original "tnpic"  program was Copyright (C) 1993-1996 Russell Marks.
    Both tnpic and gtnpic were distributed under the GNU Public License. 
*/

#if HAVE_CONFIG_H
#    include <config.h>
#endif 

#include "common.h"
#include "jpgtn.h"


struct my_error_mgr {
    struct jpeg_error_mgr pub;  /* "public" fields      */
    jmp_buf setjmp_buffer;      /* for return to caller */
};
typedef struct my_error_mgr * my_error_ptr;

static void my_error_exit(j_common_ptr cinfo)
{
    my_error_ptr myerr=(my_error_ptr) cinfo->err;
    char buf[JMSG_LENGTH_MAX];
    (*cinfo->err->format_message)(cinfo,buf);
    longjmp(myerr->setjmp_buffer, 1);
}

unsigned char *read_JPEG_file(char *filename, unsigned char **palette)
{
    FILE *infile;
    struct jpeg_decompress_struct cinfo;
    struct my_error_mgr jerr;
    int row_stride;
    unsigned char *image;
    unsigned char *pal;
    int f;
    unsigned char *ptr;

    image = NULL;

    if (((*palette) = (unsigned char *)calloc(1,768))==NULL) {
        return NULL;
    }

    pal = *palette;

    if (NULL == (infile=fopen(filename,"rb"))) {
        return NULL;
    }

    cinfo.err = jpeg_std_error(&jerr.pub);
    jerr.pub.error_exit = my_error_exit;

    if (setjmp(jerr.setjmp_buffer)) {
        jpeg_destroy_decompress(&cinfo);
        fclose(infile);
        free(pal);
        return NULL;
    }

    /* Now we can initialize the JPEG decompression object. */
    jpeg_create_decompress(&cinfo);

    jpeg_stdio_src(&cinfo,infile);

    jpeg_read_header(&cinfo,TRUE);

    /* setup parameters for decompression */
    cinfo.dct_method=JDCT_ISLOW;
    cinfo.quantize_colors=TRUE;
    cinfo.desired_number_of_colors=256;
    cinfo.two_pass_quantize=TRUE;

    /* Fix to greys if greyscale. (required to read greyscale JPEGs) */
    if (cinfo.jpeg_color_space == JCS_GRAYSCALE) {

        cinfo.out_color_space = JCS_GRAYSCALE;
        cinfo.desired_number_of_colors = 256;
        cinfo.quantize_colors = FALSE;
        cinfo.two_pass_quantize = FALSE;

        for(f=0;f<256;f++) { 
            pal[f]=pal[256+f]=pal[512+f]=f;
        }
    }

    width = cinfo.image_width;
    height = cinfo.image_height;
    image = (unsigned char *)calloc(1,width*height);

    if (image==NULL) {
        printf("Out of memory");
        longjmp(jerr.setjmp_buffer,1);
    }

    jpeg_start_decompress(&cinfo);

    /* read the palette (if greyscale, this has already been done) */
    if (cinfo.jpeg_color_space != JCS_GRAYSCALE) {

        for (f=0; f<cinfo.actual_number_of_colors; f++) {
            pal[    f]=cinfo.colormap[0][f];
            pal[256+f]=cinfo.colormap[1][f];
            pal[512+f]=cinfo.colormap[2][f];
        }
    }

    ptr = image; 
    row_stride = width;

    /* read the image */
    while (cinfo.output_scanline<height) {
        jpeg_read_scanlines(&cinfo, &ptr, 1);
        ptr += row_stride;
    }

    jpeg_finish_decompress(&cinfo);
    jpeg_destroy_decompress(&cinfo);
    fclose(infile);
    return image;
}


src/resizepic.c
/*
    Copyright (C) 2002-2012 Jeremy Madea <jpgtn@madea.net>

    This file is part of jpgtn.

    Jpgtn is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

    Jpgtn is based on two programs: tnpic (distributed with zgv) and gtnpic.
    The original "gtnpic" program was Copyright (C) 1997      Willie Daniel. 
    The original "tnpic"  program was Copyright (C) 1993-1996 Russell Marks.
    Both tnpic and gtnpic were distributed under the GNU Public License. 
*/

#if HAVE_CONFIG_H
#    include <config.h>
#endif

#include "common.h"
#include "jpgtn.h"

unsigned char *resizepic(unsigned char *image, 
                         unsigned char *palr,
                         unsigned char *palg, 
                         unsigned char *palb,
                         int currwidth, int currheight,
                         int maxsize,
                         int resize_dim)
{
    register int   a, b, x, yp, yw;
             long  y, sw, sh, lastyp;
             int   c, pixwide, pixhigh;
    unsigned char *rline;
             int   tmp, tr, tg, tb, tn;
             int   xypos;

    if (!resize_dim) {    /* The resize dimension wasn't provided... */
        if (currwidth <= currheight) { /* Picture is square or portraity */
            resize_dim = RSZ_HEIGHT;

        } else {                       /* Picture is landscapey */
            resize_dim = RSZ_WIDTH;
        }
    }

        switch(resize_dim) {
            case RSZ_HEIGHT:
                sh = maxsize;
                sw = (int)( (maxsize * ((double)currwidth)) / 
                                       ((double)currheight) );
                break;
            case RSZ_WIDTH:
                sw = maxsize;
                sh = (int)( (maxsize * ((double)currheight)) / 
                                       ((double)currwidth) );
                break;
            default:
                /* WE SHOULD NEVER GET HERE! */
                sw = sh = 0;
                break;
        }
    
    /* This is a kludge. */
    if (sw == 0) sw += 1;
    if (sh == 0) sh += 1;

    out_wide = sw; 
    out_high = sh;
    
    /* Reserve the amount of memory we will need. */
    rline = calloc(1, sw * sh * 3);
    if (rline == NULL) {
        return NULL;
    }

    for (x=0; x < (sw*sh*3); x++) {
        rline[x] = GRAY;
    }

    if (currwidth > sw) {  /* It has been reduced. */
        lastyp  = -1;
        pixhigh = (int)( ((float)currheight) / ((float)sh) + 0.5 );
        pixwide = (int)( ((float)currwidth)  / ((float)sw) + 0.5 );
        pixhigh++;
        pixwide++;
        for(y=0; y < currheight; y++) {
            yp = (y * sh) / currheight;
            if (yp != lastyp) {
                yw = (y * currwidth);

                /* we try to resample a bit. get that pentium RSN :) */
                for (x=0; x < currwidth; x++,yw++) {
                    tr = 0;
                        tg = 0;
                        tb = 0;
                        tn = 0;
                    for (b=0; (b < pixhigh) && (y+b < currheight); b++) {
                        for (a=0; (a < pixwide) && (x+a < currwidth); a++) {
                            tmp = *(image+yw+a+b*currwidth);
                            tr  += palr[tmp];
                            tg  += palg[tmp];
                            tb  += palb[tmp];
                            tn++;
                        }
                    }

                    xypos = 3*(((x*sw)/currwidth)+yp*sw);
                    tr /= tn; 
                    tg /= tn;
                    tb /= tn;
                    rline[xypos]   = tr;
                    rline[xypos+1] = tg;
                    rline[xypos+2] = tb;
                }
                lastyp = yp;
            }
        }
    } else {     /* we just leave it the same size */

        out_wide = currwidth; 
        out_high = currheight; 
        for (y=0; y < currheight; y++) {
            for (x=0; x < currwidth; x++) {
                c = image[y*currwidth+x];
                rline[3*(y*currwidth+x)]   = palr[c];
                rline[3*(y*currwidth+x)+1] = palg[c];
                rline[3*(y*currwidth+x)+2] = palb[c];
            }
        }
    }

    return rline;
}


src/writejpeg.c
/*
    Copyright (C) 2002-2012 Jeremy Madea <jpgtn@madea.net>

    This file is part of jpgtn.

    Jpgtn is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

    Jpgtn is based on two programs: tnpic (distributed with zgv) and gtnpic.
    The original "gtnpic" program was Copyright (C) 1997      Willie Daniel. 
    The original "tnpic"  program was Copyright (C) 1993-1996 Russell Marks.
    Both tnpic and gtnpic were distributed under the GNU Public License. 
*/
#if HAVE_CONFIG_H
#    include <config.h>
#endif 

#include "common.h"
#include "jpgtn.h"

/* write_JPEG_file -- Writes the global image "outimage" to a file. */

void write_JPEG_file(char *filename, int outwidth, int outheight, int quality)
{
    struct jpeg_compress_struct cinfo;
    struct jpeg_error_mgr       jerr;
    FILE     *out;
    JSAMPROW row_pointer[1];    /* pointer to JSAMPLE row[s]          */
    int      row_stride;        /* physical row width in image buffer */

    /* Step 1: Allocate and initialize JPEG compression object. */

    cinfo.err = jpeg_std_error(&jerr);
    jpeg_create_compress(&cinfo);

    /* Step 2: Specify data destination. */

        if (filename) {
            out = fopen(filename,"wb");
            if (out == NULL) {
                fprintf(stderr,"can't open %s\n",filename);
                exit(1);
            }
        } else {
            out = stdout;
        }
    jpeg_stdio_dest(&cinfo,out);

    /* Step 3: set parameters for compression */

    cinfo.image_width      = outwidth;  /* image width and height, in pix */
    cinfo.image_height     = outheight;
    cinfo.input_components = 3;         /* # of color components per pix  */
    cinfo.in_color_space   = JCS_RGB;   /* colorspace of input image      */

    jpeg_set_defaults(&cinfo);
    jpeg_set_quality(&cinfo,quality,FALSE);

    /* Step 4: Start compressor */

    jpeg_start_compress(&cinfo,TRUE);

    /* Step 5: while (scan lines remain to be written) */

    row_stride = outwidth * 3;    /* JSAMPLEs per row in image_buffer */

    while (cinfo.next_scanline < cinfo.image_height) {
        row_pointer[0] = &outimage[cinfo.next_scanline*row_stride];
        jpeg_write_scanlines(&cinfo,row_pointer,1);
    }

    /* Step 6: Finish compression */

    jpeg_finish_compress(&cinfo);
    fclose(out);

    /* Step 7: release JPEG compression object */

    jpeg_destroy_compress(&cinfo);

}


src/jpgtn.c
/*
    Copyright (C) 2002 Jeremy Madea <jeremymadea@mindspring.com>

    This file is part of jpgtn.

    Jpgtn is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

    Jpgtn is based on two programs: tnpic (distributed with zgv) and gtnpic.
    The original "gtnpic" program was Copyright (C) 1997      Willie Daniel.
    The original "tnpic"  program was Copyright (C) 1993-1996 Russell Marks.
    Both tnpic and gtnpic were distributed under the GNU Public License.
*/

#if HAVE_CONFIG_H
#    include <config.h>
#endif

#include "common.h"
#include "jpgtn.h"
#include <sys/types.h> //included to open directory
#include <dirent.h>

typedef struct node {
    char *path;
    struct node * next;
} node;

node* head = NULL;
node* cursor = NULL;


/* Private */
static void print_version(void);
static void print_usage(char *);
static int  file_exists(const char *);
static char *output_file_name(const char *,int,const char *,int,const char *);
static void get_files(char *);
static void insert_into_array(char *);



int main(int argc, char **argv)
{
    int f;                                 /* Current file to be indexed  */
    int xsiz = 0;                          /* Total output image xsize    */
    int ysiz = 0;                          /* Total output image ysize    */
    unsigned char *bigimage   = NULL;      /* Pointer to current image    */
    unsigned char *palrgb     = NULL;      /* Pointer to image palette    */
    int x=0;
    int  size = DEFAULTSIZE;               /* Size of thumbnail image.    */
    int  opt_size;
    int  opt_flag_quality   = 0;
    int  opt_flag_prefix    = 0;
    int  opt_flag_directory = 0;
    int  opt_flag_stdout    = 0;
    int  opt_flag_force     = 0;
    int  opt_flag_verbose   = 0;
    int  opt_quality;                      /* JPEG Quality                */
    int  opt_rsz_dim = RSZ_AUTO;           /* Dimension to resize. 0,1,2  */
    char *opt_directory = NULL;            /* Output directory            */
    char *opt_prefix    = NULL;            /* Output filename prefix      */
    char *imgfile       = NULL;            /* Output filename             */

    int grid_mode = 0;
    char *source_directory = NULL;
    int grid_width = 0;
    int grid_height = 0;
    int src_directory = 0;
    int rows_per_image = 0;
    int columns_per_image = 0;



    /* Check command args */
    if (argc < 2) {
        print_usage(argv[0]);
        exit(EXIT_FAILURE);
    }

    while (-1 != (x = getopt(argc,argv,"Vh?SHGWfvs:q:d:p:R:C:J:K:D:"))) {
        switch(x) {
            case 'V':
                print_version();
                exit(EXIT_FAILURE);
                break;
            case 'h':
                print_usage(argv[0]);
                exit(EXIT_FAILURE);
                break;
            case 'q':
                sscanf(optarg,"%d",&opt_quality);
                if(opt_quality < 1 || opt_quality > 100) {
                    opt_quality = 70;
                }
                opt_flag_quality = 1;
                break;
            case 's':
                sscanf(optarg,"%d",&opt_size);
                if ((opt_size < MINSIZE) || (opt_size > MAXSIZE)) {
                    opt_size = DEFAULTSIZE;
                }
                size = opt_size;
                break;
            case 'f':
                opt_flag_force = 1;
                break;
            case 'v':
                opt_flag_verbose++;
                break;
            case 'S':
                opt_flag_stdout = 1;
                break;
            case 'H':
                opt_rsz_dim = RSZ_HEIGHT;
                break;
            case 'w':
                opt_rsz_dim = RSZ_WIDTH;
                break;
            case 'd':
                opt_directory = optarg;
                opt_flag_directory = 1;
                break;
            case 'p':
                opt_prefix = optarg;
                opt_flag_prefix = 1;
                break;
            case 'D':
                source_directory = optarg;
                src_directory = 1;
                break;
            case 'R':
                sscanf(optarg,"%i",&rows_per_image);
                break;
            case 'C':
                sscanf(optarg,"%i",&columns_per_image);
                break;
            case 'J':
                sscanf(optarg,"%i",&grid_height);
                break;
            case 'K':
                sscanf(optarg,"%i",&grid_width);
                break;
            case 'G':
                grid_mode = 1;
                break;

            case '?':
                print_usage(argv[0]);
                exit(EXIT_FAILURE);
                break;
            default:
                fprintf(stderr,"Unknown error. Exiting.\n");
                exit(EXIT_FAILURE);
                break;
            }
        }

    if ((!opt_flag_stdout) && (!opt_flag_directory) && (!opt_flag_prefix)) {
        opt_flag_prefix = 1;
        opt_prefix = "tn_";
    }

    if (!opt_flag_quality) opt_quality = 70;

    if (opt_flag_verbose > 2) {
        printf("quality: %d\n",opt_quality);
    }

    if(grid_mode == 1) {
        //grid mode enabled
        //TODO: check parameters existence

        //get images
        if(src_directory == 1) {

            get_files(source_directory);
            for(node* i = head; i != NULL; i = i->next ){
                printf("%s\n", i->path);
            }
        }



    } else {
        if (argc - optind < 1) {
            print_usage(argv[0]);
            exit(EXIT_FAILURE);
        }
        /* The big loop */
        for (f = optind; f < argc; f++) {

            if (!opt_flag_stdout) {
                imgfile = output_file_name(argv[f],
                                           opt_flag_directory, opt_directory,
                                           opt_flag_prefix, opt_prefix);
                if (imgfile != NULL) {
                    if (!opt_flag_force && file_exists(imgfile)) {
                        fprintf(stderr,"Skipping %s: File exists\n", imgfile);
                        continue;
                    }
                } else {
                    fprintf(stderr, "Could not allocate space for filename!\n");
                    exit(EXIT_FAILURE);
                }
            }

            /* Get a pointer to the big image or NULL on error */
            bigimage=read_JPEG_file(argv[f], &palette);
            if (NULL == bigimage) {
                fprintf(stderr, "read_JPEG_file: Error\n");
                if(outimage) free(outimage);
                exit(EXIT_FAILURE);
            }
            palrgb=palette;

            /* Returns a pointer to the resized image or NULL on error */
            outimage = resizepic(bigimage, palrgb, palrgb+256, palrgb+512,
                                 width, height, size, opt_rsz_dim);
            xsiz = out_wide;
            ysiz = out_high;

            if(outimage == NULL) {
                fprintf(stderr,"outimage = NULL ???\n");
                if(bigimage) free(bigimage);
                exit(EXIT_FAILURE);
            }

            if ((xsiz < MINSIZE) && (ysiz < MINSIZE)) {
                 fprintf(stderr, "thumbnail too small...\n");
            }

            free(palrgb);
            free(bigimage);

            if (! opt_flag_stdout) {
                if (opt_flag_verbose) {
                    printf("%s",imgfile);
                    if (opt_flag_verbose > 1) printf(" %dx%d",xsiz,ysiz);
                    printf("\n");
                }
                write_JPEG_file(imgfile,xsiz,ysiz,opt_quality);
                free(imgfile);
                imgfile = NULL;
                free(outimage);

            } else {
                write_JPEG_file(NULL,xsiz,ysiz,opt_quality);
                exit(EXIT_SUCCESS);
            }
        }
    }




    exit (EXIT_SUCCESS);
}

static int file_exists(const char *path) {
    struct stat buf;

    /* Crude check. If we can stat it successfully, we assume it exists. */
    /* Otherwise we return zero, meaning that this function can't tell. */

    if (0 == stat(path, &buf)) {
        return(1);
    } else {
        return(0);
    }
}

static char *output_file_name(const char *infile,
                              int dflag, const char *dir,
                              int pflag, const char *pre)
{
    int   len       = 0;
    char *infiledup = NULL;
    char *ifn       = NULL;
    char *name      = NULL;

    /* Make a copy of the input file name. */
    infiledup = strdup(infile);
    if (! infiledup) {
        return(NULL);
    }

    /* Ignore path information at the beginning of the filename. */
    ifn = strrchr(infiledup,'/');
    if (!ifn) {
        ifn = infiledup;
    } else {
        ifn++;
    }

    /* Estimate max length we'll need; add two for "./" and one for null. */
    len = strlen(ifn) + 3;
    if (dir) len += strlen(dir);
    if (pre) len += strlen(pre);

    /* Reserve space for it... */
    name = (char *)malloc(len);

    if (name == NULL) {
        return NULL;
    }

    *name = '\0';

    /* Construct the output file name. */
    if (dflag) {
       strcpy(name, dir);
       len = strlen(dir);

        /* Add "/" on dir if needed.*/
        if (dir[len - 1] != '/') {
            strcat(name,"/");
        }
    } else {
        /* strcpy(name, "./"); */
    }

    if (pflag) {
        strcat(name, pre);
        strcat(name, ifn);
    } else {
        strcat(name, ifn);
    }

    free(infiledup);
    return(name);
}

static void print_usage(char *prog)
{
    print_version();
    printf("usage: %s [switches] [inputfiles]\n",prog);
    printf("  -d directory  Set the output directory.\n");
    printf("  -f            Force overwrite of existing output files.\n");
    printf("  -h            Display this help message and exit.\n");
    printf("  -p prefix     Set the output filename prefix.\n");
    printf("  -q quality    Set the output quality.\n");
    printf("  -s size       Set the size of the longest output dimension. ");
    printf(" (See -H and -V)\n");
    printf("  -v            Verbosity level. (May be used multiple times.)\n");
    printf("  -H            Make the -s switch refer to the output height.\n");
    printf("  -S            Process only one file and output to stdout.\n");
    printf("  -W            Make the -s switch refer to the output width.\n");
    printf("  -G            Set the grid mode.\n");
    printf("  -D directory  Set the source directory.\n");
    printf("  -R            Set the number of rows in the grid.\n");
    printf("  -C            Set the number of columns in the grid.\n");
    printf("  -J            Set the height of the grid image.\n");
    printf("  -K            Set the width of the grid image.\n");
    printf("  -V            Display version and exit.\n");
    printf("\nSee jpgtn(1) for more details.\n\n");
}

static void print_version(void)
{
    printf("jpgtn : Version %s\n",VERSION);
    printf("Copyright (c) 2002 Jeremy Madea <jeremymadea@mindspring.com>\n\n");
}

static void get_files(char *folder_name) {
    DIR           *d;
    struct dirent *dir;
    //check if the path provided has a slash
    if( folder_name[strlen(folder_name) - 1] != '/') {
        strcat(folder_name, "/");
    }
    d = opendir(folder_name);
    if (d) {
        while ((dir = readdir(d)) != NULL)
        {
            char file_path[strlen(folder_name) +  strlen(dir->d_name) + 2];
            if( strcmp(dir->d_name, ".") == 0 || strcmp(dir->d_name, "..") == 0 ) {
                continue;
            }
            //build the path
            strcpy( file_path, folder_name);
            strcat( file_path, dir->d_name);
            //check if file is a folder
            if( opendir(file_path) ){
                //call recursively this function
                get_files(file_path);
            } else {
                insert_into_array(file_path);
            }

        }

        closedir(d);
    }

}

static void insert_into_array(char *str){
    char *node_str = (char *) malloc(strlen(str) +1);
    strcpy(node_str, str);
    node* new = malloc(sizeof(node));
    new->path = node_str;
    new->next = NULL;
    if( head == NULL){
        head = new;
        cursor = head;
    } else{
        cursor->next = new;
        cursor = new;
    }
}
